### PR TITLE
Improve partition cache

### PR DESCRIPTION
**Pull request by hubcio**
First version of `safer` cache, started using VecDeque.
Every write to cache is accounted via atomic counter.
Counter is shared between all partitions, so when summing
all messages, we get an rought estimate of used memory.
    
When cache is exceeded, messages are proportionally
removed from each partiton cache.
    
According to flamegraph, `clean_cache` should take around
2.37% of CPU time, so it's not bad. Some optimizations
were introduced in this commit, so the change shouldn't
be that severe in terms of performance.
    
Besides that, added module for parsing memory resource quota
along with serialization/deserialization, so we can store
it directly in the config and seamlessly use it.

### CODE REVIEW

**Comment by hubcio** [comment]
This is just for demonstration ;) will polish PR later on.

**Comment by hubcio** [comment]
Still working on the issue. It turns out it's hard to write good, performant code.

**Comment by hubcio** [comment]
I encourage you guys to review @spetz @BartoszCiesla @daniel-pape 🦀

**Comment by hubcio** [comment]
Shall I move clean_cache method call to `channel.rs` and handle it there?

File: server/src/streaming/segments/messages.rs [hunk 1]
```diff
@@ -155,33 +155,77 @@ impl Segment {
...
+        if let Some(time_indexes) = &mut self.time_indexes {
+            time_indexes.reserve(len);
         }
 
-        self.current_size_bytes += message.get_size_bytes();
-        self.current_offset = message.offset;
-        self.unsaved_messages.as_mut().unwrap().push(message);
+        // Not the prettiest code. It's done this way to avoid repeatably
+        // checking if indexes and time_indexes are Some or None.
+        if self.indexes.is_some() && self.time_indexes.is_some() {
```
Comment (spetz) [inline server/src/streaming/segments/messages.rs:178]: Maybe declare the mutable `indexes` and `time_indexes` variables, to make code a bit more readable?

```rust
if self.indexes.is_some() && self.time_indexes.is_some() {
    let indexes = self.indexes.as_mut().unwrap();
    let time_indexes = self.time_indexes.as_mut().unwrap();
    for message in messages {
        let relative_offset = (message.offset - self.start_offset) as u32;

        indexes.push(Index {
            relative_offset,
            position: self.current_size_bytes,
        });

        time_indexes.push(TimeIndex {
            relative_offset,
            timestamp: message.timestamp,
        });

        self.current_size_bytes += message.get_size_bytes();
        self.current_offset = message.offset;
        unsaved_messages.push(message.clone());
    }
} else if self.indexes.is_some() {
    let indexes = self.indexes.as_mut().unwrap();
    for message in messages {
        let relative_offset = (message.offset - self.start_offset) as u32;

        indexes.push(Index {
            relative_offset,
            position: self.current_size_bytes,
        });

        self.current_size_bytes += message.get_size_bytes();
        self.current_offset = message.offset;
        unsaved_messages.push(message.clone());
    }
} else if self.time_indexes.is_some() {
    let time_indexes = self.time_indexes.as_mut().unwrap();
    for message in messages {
        let relative_offset = (message.offset - self.start_offset) as u32;

        time_indexes.push(TimeIndex {
            relative_offset,
            timestamp: message.timestamp,
        });

        self.current_size_bytes += message.get_size_bytes();
        self.current_offset = message.offset;
        unsaved_messages.push(message.clone());
    }
} else {
    for message in messages {
        self.current_size_bytes += message.get_size_bytes();
        self.current_offset = message.offset;
        unsaved_messages.push(message.clone());
    }
}
  ```

**Review by spetz — APPROVED** [review]
Well done, this is a very important feature, that will greatly improve the server reliability.

File: server/src/streaming/segments/messages.rs (hunk 1 above)
Comment (hubcio) [reply server/src/streaming/segments/messages.rs:178]: In next, refactor commit I will do it.

**Pull request merged**